gapil/compiler/statements.go

```diff
@@ -346,11 +346,15 @@ func (c *compiler) read(s *scope, n *semantic.Read) {
 }
 
 func (c *compiler) return_(s *scope, n *semantic.Return) {
-	val := c.expression(s, n.Value)
-	c.reference(s, val, n.Value.ExpressionType())
+	if n.Value != nil {
+		val := c.expression(s, n.Value)
+		c.reference(s, val, n.Value.ExpressionType())
 
-	retTy := c.returnType(c.currentFunc) // <error, value>
-	c.doReturn(s, s.Zero(retTy).Insert(retValue, val))
+		retTy := c.returnType(c.currentFunc) // <error, value>
+		c.doReturn(s, s.Zero(retTy).Insert(retValue, val))
+	} else {
+		c.doReturn(s, nil)
```

Comment:
Relying here on `<0, 0>` being an acceptable return value, when these aren't legal values for all types.

The `string`, `class` and `map` type's `initialValue`s  are not the same as an `codegen.Builder.Zero()` values. Returning a zero value for these will likely cause NPEs for these types as the compiler expects values to be non-null.

We probably want:

```go
var val *codegen.Value
if n.Value != nil {
	val = c.expression(s, n.Value)
} else {
	val = c.initialValue(s, n.Value.ExpressionType())
}
c.reference(s, val, n.Value.ExpressionType())
retTy := c.returnType(c.currentFunc) // <error, value>
ret := s.Zero(retTy).Insert(retValue, val)
c.doReturn(s, ret)
```

Note: I also made the same mistake in the proto branch.

Reply:
Done

Reply:
`val = c.initialValue(s, n.Value.ExpressionType())` Won't work, but I have an update that should fix it.